### ISSUE TITLE
"Backup all fittings" hangs on version 2.20.2 using savedddata.db from previous versions

### ISSUE THREAD

**Issue by mmikelst** [pyfa-org/Pyfa#2173]
### **Other relevant information:**
Attempting to import .xml files exported from previous versions of pyfa into 2.20.2 ALSO fails with an error message (message lost during testing). To reproduce, try to import a set of all fittings backed up (the .xml file) from a previous version of pyfa (confirmed in 2.17.2) into pyfa version 2.20.2.

**Comment by DarkFenX** [comment]
```
Exception in thread Thread-6:
Traceback (most recent call last):
  File "/usr/lib/python3.8/threading.py", line 932, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.8/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "/home/dfx/Workspace/eve/pyfa/service/port/port.py", line 83, in backupFitsWorkerFunc
    backedUpFits = Port.exportXml(svcFit.getInstance().getAllFits(), iportuser)
  File "/home/dfx/Workspace/eve/pyfa/service/fit.py", line 104, in getAllFits
    fits = eos.db.getFitList()
  File "/home/dfx/Workspace/eve/pyfa/eos/db/saveddata/queries.py", line 324, in getFitList
    fits = removeInvalid(saveddata_session.query(Fit).options(*eager).all())
  File "/usr/lib/python3/dist-packages/sqlalchemy/orm/query.py", line 3244, in all
    return list(self)
  File "/usr/lib/python3/dist-packages/sqlalchemy/orm/loading.py", line 101, in instances
    cursor.close()
  File "/usr/lib/python3/dist-packages/sqlalchemy/util/langhelpers.py", line 68, in __exit__
    compat.raise_(
  File "/usr/lib/python3/dist-packages/sqlalchemy/util/compat.py", line 178, in raise_
    raise exception
  File "/usr/lib/python3/dist-packages/sqlalchemy/orm/loading.py", line 81, in instances
    rows = [proc(row) for row in fetch]
  File "/usr/lib/python3/dist-packages/sqlalchemy/orm/loading.py", line 81, in <listcomp>
    rows = [proc(row) for row in fetch]
  File "/usr/lib/python3/dist-packages/sqlalchemy/orm/loading.py", line 599, in _instance
    state.manager.dispatch.load(state, context)
  File "/usr/lib/python3/dist-packages/sqlalchemy/event/attr.py", line 322, in __call__
    fn(*args, **kw)
  File "/usr/lib/python3/dist-packages/sqlalchemy/orm/mapper.py", line 3353, in _event_on_load
    instrumenting_mapper._reconstructor(state.obj())
  File "/home/dfx/Workspace/eve/pyfa/eos/saveddata/fit.py", line 111, in init
    self.__ship = Ship(item, self)
  File "/home/dfx/Workspace/eve/pyfa/eos/saveddata/ship.py", line 53, in __init__
    self.__itemModifiedAttributes.original = dict(self.item.attributes)
  File "/home/dfx/Workspace/eve/pyfa/eos/gamedata.py", line 268, in attributes
    self.moveAttrs()
  File "/home/dfx/Workspace/eve/pyfa/eos/gamedata.py", line 238, in moveAttrs
    self.__attributes[info.name] = attr
  File "/usr/lib/python3/dist-packages/sqlalchemy/orm/collections.py", line 1281, in __setitem__
    value = __set(self, value, _sa_initiator)
  File "/usr/lib/python3/dist-packages/sqlalchemy/orm/collections.py", line 1078, in __set
    item = executor.fire_append_event(item, _sa_initiator)
  File "/usr/lib/python3/dist-packages/sqlalchemy/orm/collections.py", line 714, in fire_append_event
    return self.attr.fire_append_event(
  File "/usr/lib/python3/dist-packages/sqlalchemy/orm/attributes.py", line 1176, in fire_append_event
    value = fn(state, value, initiator or self._append_token)
  File "/usr/lib/python3/dist-packages/sqlalchemy/orm/unitofwork.py", line 51, in append
    sess._save_or_update_state(item_state)
  File "/usr/lib/python3/dist-packages/sqlalchemy/orm/session.py", line 2000, in _save_or_update_state
    self._save_or_update_impl(st_)
  File "/usr/lib/python3/dist-packages/sqlalchemy/orm/session.py", line 2346, in _save_or_update_impl
    self._update_impl(state)
  File "/usr/lib/python3/dist-packages/sqlalchemy/orm/session.py", line 2329, in _update_impl
    to_attach = self._before_attach(state, obj)
  File "/usr/lib/python3/dist-packages/sqlalchemy/orm/session.py", line 2414, in _before_attach
    raise sa_exc.InvalidRequestError(
sqlalchemy.exc.InvalidRequestError: Object '<AttributeInfo at 0x7f725f6c98b0>' is already attached to session '1' (this is '4')
```

It most likely shouldn't work at all.

**Comment by DarkFenX** [comment]
Same for HTML export actually

**Comment by thespus** [comment]
Looks like the same thing when exporting an individual fit to XML from the Fit pull down  menu.

---------------------------------------------------------

pyfa v2.20.2
EVE Data Version: 1706308 (2020-04-15 04:32:35)

OS version: Windows-10-10.0.18362-SP0
Python version: 3.7.7 (tags/v3.7.7:d7c567b08f, Mar 10 2020, 10:41:24) [MSC v.1900 64 bit (AMD64)]
wxPython version: 4.0.6 (wxWidgets 3.0.5)
SQLAlchemy version: 1.3.16
Logbook version: 1.5.3
Requests version: 2.23.0
Dateutil version: 2.8.1

####################

Traceback (most recent call last):
  File "gui\builtinShipBrowser\sfBrowserItem.py", line 354, in OnLeftDown
AttributeError: 'bool' object has no attribute 'GetState'

**Comment by DarkFenX** [comment]
@thespus this is a separate issue. Please create a new one, specifying the fit you want to export. I cannot reproduce it.

**Comment by DarkFenX** [comment]
The functionality was broken by fix for #2167. The fix was alright, but the way we worked with sqlalchemy before the fix is awful, no thoughts about session scoping and thread safety, so have to fix loading of fits (with items) in another thread separately.

**Comment by DarkFenX** [comment]
Issue is that we now use scoped sqlalchemy sessions, one session for each thread. Here, it loads fits in the export thread session, and part of loading fit is loading its items. When it loads items, it tries to move item basic attributes (mass, volume, capacity) from on-item descriptors to attribute storage. To avoid messing with sqlalchemy concurrency stuff, I would rather move this logic to db composition script.

**Comment by DarkFenX** [comment]
Fixed in 21ea9ce579f5ad693e5cdd4d815a4034d4ac4929
